### ISSUE TITLE
How to Save single Optimization Experiment

### ISSUE THREAD

**Issue by StanleyYoo** [facebook/Ax#1780]
Hi,
I have followed [Tutorial](https://ax.dev/tutorials/gpei_hartmann_developer.html) and tried to save the experiment by JSON but following error occurs: ax.exceptions.storage.JSONEncodeError: Object <__main__.MyRunner object at 0x0000016ED79986A0> passed to `object_to_json` (of type <class '__main__.MyRunner'>, module: __main__) is not registered with a corresponding encoder in ENCODER_REGISTRY.

I have updated MyRunner class to save data as a record:
```
class MyRunner(Runner):
    def run(self, trial):
        # trial_metadata = {"name": str(trial.index)}
        # name_to_params = {arm.: arm.params for arm in trial.a}
        for arm_name, arm in trial.arms_by_name.items():
            records = []
            params = arm.parameters
            records.append(
                {
                    "arm_name": arm_name,
                    "x0": params["x0"],
                    "x1": params["x1"],
                    "x2": params["x2"],
                    "x3": params["x3"],
                    "x4": params["x4"],
                    "x5": params["x5"],
                    "trial_index": trial.index,
                }
            )
            df = pd.DataFrame.from_records(records)
        return df
```
Could you please let me know what causes an error?

**Comment by bernardbeckerman** [comment]
Just popping in to say this is so cool!

**Comment by bernardbeckerman** [comment]
@StanleyYoo thanks for raising this here! 

It seems like the error you received is indicating that the `MyRunner` class you're trying to save (i.e., the one attached to the experiment) is not the same as the `MyRunner` class in the registry (i.e., the MyRunner present in your notebook at the time of saving). I believe if you replace the previous MyRunner with your new MyRunner during runner definition [here](https://ax.dev/tutorials/gpei_hartmann_developer.html#3.-Define-a-Runner), and then re-run the notebook from the beginning, things should work.

As a quick follow-up, it seems like your runner may be duplicating some data that is already available elsewhere on the trial, and I believe some added context on what you're trying to accomplish may help me guide you to a better solution. For instance, perhaps [`exp_to_df`](https://ax.dev/api/service.html#ax.service.utils.report_utils.exp_to_df) could be helpful here, e.g., if you try

```
from ax.service.utils.report_utils import exp_to_df
exp_to_df(exp)
```

you may get the dataframe you're trying to create, but without having to alter the runner and pass around extraneous data. Let me know if that helps!

**Comment by StanleyYoo** [comment]
Hi @bernardbeckerman! Thanks for your comments indeed. 
Your suggestion resolved my concern of saving the outcomes more efficient way. Still my initial intention to use [save_experiment](https://ax.dev/docs/storage.html) is to restore the experiment after some numbers of GPEI and re-train with further trial once necessity arises. Still the same error published with following code even though ### **MyRunner** class has not been modified. Do you have any idea?
`from pathlib import Path

import plotly.io as pio
from ax import (
    ChoiceParameter,
    ComparisonOp,
    Experiment,
    FixedParameter,
    Metric,
    Objective,
    OptimizationConfig,
    OrderConstraint,
    OutcomeConstraint,
    ParameterType,
    RangeParameter,
    Runner,
    SearchSpace,
    SumConstraint,
)
from ax.metrics.hartmann6 import Hartmann6Metric
from ax.metrics.l2norm import L2NormMetric
from ax.modelbridge.registry import Models
from ax.storage.json_store.save import save_experiment
from ax.utils.notebook.plotting import init_notebook_plotting, render

init_notebook_plotting()


pio.renderers.default = "browser"
cwd = Path().cwd()
filepathString = cwd / "simOut.json"
filepath = Path(filepathString)
last_part = filepath.parts[-1]
print(last_part)
print(last_part.endswith('ending'))
print(filepath)
filepath = "C:/Users/seuyoo/Development/mvx/tests/demos/simulink_developer/simOut.json"
# exit(0)
hartmann_search_space = SearchSpace(
    parameters=[
        RangeParameter(
            name=f"x{i}", parameter_type=ParameterType.FLOAT, lower=0.0, upper=1.0
        )
        for i in range(6)
    ]
)

sum_constraint = SumConstraint(
    parameters=[
        hartmann_search_space.parameters["x0"],
        hartmann_search_space.parameters["x1"],
    ],
    is_upper_bound=True,
    bound=5.0,
)

order_constraint = OrderConstraint(
    lower_parameter=hartmann_search_space.parameters["x0"],
    upper_parameter=hartmann_search_space.parameters["x1"],
)

param_names = [f"x{i}" for i in range(6)]
optimization_config = OptimizationConfig(
    objective=Objective(
        metric=Hartmann6Metric(name="hartmann6", param_names=param_names),
        minimize=True,
    ),
    outcome_constraints=[
        OutcomeConstraint(
            metric=L2NormMetric(name="l2norm", param_names=param_names, noise_sd=0.2),
            op=ComparisonOp.LEQ,
            bound=1.25,
            relative=False,
        )
    ],
)



class MyRunner(Runner):
    def run(self, trial):
        trial_metadata = {"name": str(trial.index)}
        return trial_metadata

exp = Experiment(
    name="test_hartmann",
    search_space=hartmann_search_space,
    optimization_config=optimization_config,
    runner=MyRunner(),
)


NUM_SOBOL_TRIALS = 5
NUM_BOTORCH_TRIALS = 15

print(f"Running Sobol initialization trials...")
sobol = Models.SOBOL(search_space=exp.search_space)

for i in range(NUM_SOBOL_TRIALS):
    # Produce a GeneratorRun from the model, which contains proposed arm(s) and other metadata
    generator_run = sobol.gen(n=1)
    # Add generator run to a trial to make it part of the experiment and evaluate arm(s) in it
    trial = exp.new_trial(generator_run=generator_run)
    # Start trial run to evaluate arm(s) in the trial
    trial.run()
    # Mark trial as completed to record when a trial run is completed
    # and enable fetching of data for metrics on the experiment
    # (by default, trials must be completed before metrics can fetch their data,
    # unless a metric is explicitly configured otherwise)
    trial.mark_completed()

for i in range(NUM_BOTORCH_TRIALS):
    print(
        f"Running GP+EI optimization trial {i + NUM_SOBOL_TRIALS + 1}/{NUM_SOBOL_TRIALS + NUM_BOTORCH_TRIALS}..."
    )
    # Reinitialize GP+EI model at each step with updated data.
    gpei = Models.BOTORCH(experiment=exp, data=exp.fetch_data())
    generator_run = gpei.gen(n=1)
    trial = exp.new_trial(generator_run=generator_run)
    trial.run()
    trial.mark_completed()
    save_experiment(exp, filepath)

print("Done!")`

**Comment by StanleyYoo** [comment]
Hi @bernardbeckerman ! Thanks for your comments indeed.
Your suggestion resolved my concern of saving the outcomes more efficient way. Still my initial intention to use [save_experiment](https://ax.dev/docs/storage.html) is to restore the experiment after some numbers of GPEI and re-train with further trial once necessity arises. Still the same error published with following code even though MyRunner class has not been modified. Do you have any idea?

`for i in range(NUM_SOBOL_TRIALS):
    # Produce a GeneratorRun from the model, which contains proposed arm(s) and other metadata
    generator_run = sobol.gen(n=1)
    # Add generator run to a trial to make it part of the experiment and evaluate arm(s) in it
    trial = exp.new_trial(generator_run=generator_run)
    # Start trial run to evaluate arm(s) in the trial
    trial.run()
    # Mark trial as completed to record when a trial run is completed
    # and enable fetching of data for metrics on the experiment
    # (by default, trials must be completed before metrics can fetch their data,
    # unless a metric is explicitly configured otherwise)
    trial.mark_completed()

for i in range(NUM_BOTORCH_TRIALS):
    print(
        f"Running GP+EI optimization trial {i + NUM_SOBOL_TRIALS + 1}/{NUM_SOBOL_TRIALS + NUM_BOTORCH_TRIALS}..."
    )
    # Reinitialize GP+EI model at each step with updated data.
    gpei = Models.BOTORCH(experiment=exp, data=exp.fetch_data())
    generator_run = gpei.gen(n=1)
    trial = exp.new_trial(generator_run=generator_run)
    trial.run()
    trial.mark_completed()
    save_experiment(exp, filepath)`

**Comment by StanleyYoo** [comment]
> ssss

**Comment by StanleyYoo** [comment]
`dddd`

**Comment by bernardbeckerman** [comment]
It seems like you're not including the encoder per [section 9 of the tutorial](https://ax.dev/tutorials/gpei_hartmann_developer.html#9.-Save-to-JSON-or-SQL). If instead of

```
save_experiment(exp, filepath)
```

I use

```
save_experiment(exp, filepath, encoder_registry=bundle.encoder_registry)
```

this succeeds. Closing this out but please feel free to open a new issue if you need more assistance : )
